### PR TITLE
rewrite multipart parsing

### PR DESCRIPTION
**Pull request by adamhigerd**
### Intent

Backports a fix from rstudio-pro. The MIME multipart parsing had a number of flaws, so it got replaced with one that's more spec-compliant.

### Approach

Basically just a full rewrite.

### Automated Tests

Properly covered by unit tests.

### Documentation

Internal bugfix, N/A. News file was already updated in pro.

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md`
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [x] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [x] This PR passes all local unit tests

### CODE REVIEW

**Review by cm421 — APPROVED** [review]

File: src/cpp/core/http/Util.cpp [hunk 1]
```diff
@@ -121,117 +125,335 @@ void buildQueryString(const Fields& fields, std::string* pQueryString)
...
    }
-   
-   // extract the fields
-   size_t beginBoundaryLoc = body.find(boundary);
-   size_t endBoundaryLoc = body.find("\r\n" + boundary,
-                                     beginBoundaryLoc+boundary.size());
-   while (endBoundaryLoc != std::string::npos)
+
+   // Per RFC 1341, multipart-body ends immediately after the `--` with no CRLF necessary.
+   size_t terminatorPos = body.find("\r\n--" + boundary + "--");
```
Comment (kevinushey) [inline src/cpp/core/http/Util.cpp:387]: **Close-delimiter pre-scan can truncate or drop parts (confirmed by run).**

`body.find("\r\n--" + boundary + "--")` matches the first byte-level occurrence anywhere in the body, which corrupts or drops parts in three shapes I reproduced against a harness of this parser:

1. Empty final part `content\r\n--BB\r\n--BB--` matches at the delimiter's shared CRLF, leaving an orphaned `\r\n--BB` the `BoundaryFinder` can't match, so the field/file contents become `content\r\n--BB` (old parser: `content`).
2. A preamble line `\r\n--b--junk` at position 0 makes `terminatorPos == 0`, and the `if (!terminatorPos)` guard below aborts the whole parse -- all fields dropped (old parser parsed them).
3. `\r\n--boundary--` inside a part's content truncates the view and silently drops every later part.

The `BoundaryFinder`'s `NeedTerminatorDash2` state already handles the terminator correctly -- in the harness the output is byte-identical without this pre-scan. Consider letting the finder be the single terminator authority rather than pre-scanning here.

File: src/cpp/core/http/Util.cpp [hunk 2]
```diff
@@ -121,117 +125,335 @@ void buildQueryString(const Fields& fields, std::string* pQueryString)
...
+            std::string valueStr;
+            valueStr.reserve(value.size());
+            bool escape = false;
+            for (char ch : value)
+            {
+               if (escape)
+               {
+                  escape = false;
+               }
+               else if (ch == '\\')
```
Comment (kevinushey) [inline src/cpp/core/http/Util.cpp:349]: **Quoted-param backslash-unescaping mangles real filenames (confirmed by run).**

WHATWG-conformant browsers (Chrome/Firefox) percent-encode `\"` and send literal backslashes raw in `filename`/`name`, so unescaping `\\` here mangles legitimate uploads -- a regression versus the old regex, which preserved the raw bytes. RFC 7578 s4.2 explicitly warns against relying on quoted-string escaping for these values.

Reproduced: `filename="C:\\Users\\me\\f.txt"` yields `uploadedFile.name == "C:Usersmef.txt"`; a file literally named `a\\b.txt` becomes `ab.txt`; field *names* containing backslashes likewise stop matching the key the server looks up via `formFieldValue()`.

File: src/cpp/core/http/Util.cpp [hunk 3]
```diff
@@ -121,117 +125,335 @@ void buildQueryString(const Fields& fields, std::string* pQueryString)
...
+                  if (matchPos == boundary.size())
+                     state = NeedTrailingCR;
+               }
+               else
+                  state = NeedLeadingCR;
+               break;
+            case NeedTrailingCR:
+               if (ch == '\r')
+                  state = NeedTrailingLF;
+               else if (ch == '-')
```
Comment (kevinushey) [inline src/cpp/core/http/Util.cpp]: **Padded `--boundary <spaces>--` accepted as a close-delimiter (confirmed by run).**

This `else if (ch == '-')` is reachable after transport padding has been consumed, so `--boundary   --` is treated as a close-delimiter and `return Range(matchStart, end)` consumes the remainder. RFC 2046 only allows linear whitespace *after* the closing `--`, not before it. The `find()` pre-scan at line 361 doesn't recognize the padded form, so the two terminator grammars disagree.

Reproduced: a part `f1` containing a line `--BB --x`, then a real part `f2` and a real terminator -> new parser truncates `f1` and drops `f2` entirely (old parser returned both). Also reachable via the segment-start fast path (a part whose first header line is `--BB  --junk` zeroes out the whole form). Requires boundary-collision-violating input, but the failure is silent whole-form data loss.

File: src/cpp/core/http/Util.cpp [hunk 4]
```diff
@@ -121,117 +125,335 @@ void buildQueryString(const Fields& fields, std::string* pQueryString)
...
-      size_t beginPart = beginBoundaryLoc + boundary.size();
-      size_t partLength = endBoundaryLoc - beginPart;
-      std::istringstream partStream(body.substr(beginPart, partLength));
+      // No sections, just a terminating boundary
+      LOG_WARNING_MESSAGE("Invalid multipart/form-data: no sections");
+      return;
+   }
+   // Be permissinve beyond the strict requirements of RFC 1341:
+   // Use best effort to read the last part even if the terminator is missing.
+   if (terminatorPos == std::string::npos)
```
Comment (kevinushey) [inline src/cpp/core/http/Util.cpp:396]: **Missing-terminator fallback leaks trailing bytes into file contents (confirmed by run).**

The best-effort path (`terminatorPos = body.size()`) includes the last part's trailing CRLF -- and, when the body ends with a dangling `\r\n--boundary` fragment, that whole fragment -- in the extracted value. `trim()` rescues text fields, but uploaded *file* contents are shipped contaminated (the old code parsed nothing on this path).

Reproduced: body ending `...PAYLOAD\r\n` (no terminator) -> `uploadedFile.contents == "PAYLOAD\r\n"` (2 spurious bytes, breaks checksum validation); body ending `...content\r\n--BB` (truncated mid-delimiter) -> field value `content\r\n--BB`. Consider stripping the trailing CRLF/delimiter fragment, or not extending the new leniency to file parts.

File: src/cpp/core/http/Util.cpp [hunk 5]
```diff
@@ -121,117 +125,335 @@ void buildQueryString(const Fields& fields, std::string* pQueryString)
...
+
+   // iterate over the multipart sections
+   BoundaryFinder finder(boundary);
+   auto iter = ba::make_split_iterator(multipart, finder);
+   auto endIter = ba::split_iterator<std::string_view::const_iterator>();
+   bool isPreamble = true;
+   for (; iter != endIter; ++iter)
+   {
+      auto partRange = *iter;
+      if (isPreamble)
```
Comment (kevinushey) [inline src/cpp/core/http/Util.cpp:409]: **Non-CRLF preamble silently drops the first part, with no warning (confirmed by run).**

Any junk before the first delimiter that isn't CRLF-terminated (bare-LF line ending, BOM, proxy garbage) makes the strict finder classify the entire first part as preamble and discard it -- the old code's `body.find(boundary)` parsed it. This is the likeliest real-world failure mode, and it's the only one with no `LOG_WARNING` (the missing-boundary and no-sections cases both warn).

Reproduced: `junk\n--BB\r\n<part a>\r\n--BB\r\n<part b>\r\n--BB--` -> only field `b` is returned; part `a` vanishes with zero diagnostics while the request still completes successfully. At minimum, log when the skipped preamble segment is non-empty so a "first upload field missing" report is debuggable.

File: src/cpp/core/http/Util.cpp [hunk 6]
```diff
@@ -121,117 +125,335 @@ void buildQueryString(const Fields& fields, std::string* pQueryString)
...
+         {
+            escape = false;
+         }
+         else if (*iter == '\\')
+         {
+            escape = true;
+         }
+         else if (*iter == '"')
+         {
+            inQuotes = !inQuotes;
```
Comment (kevinushey) [inline src/cpp/core/http/Util.cpp:168]: **Stray quote in an unquoted token swallows later params (confirmed by run).**

`SemiFinder` toggles `inQuotes` on every `"`, including inside unquoted tokens, so a stray or unterminated quote absorbs the following `;` separators and later params (e.g. `filename=`) into the previous value. The upload is silently demoted to a text field where the old parser produced a File -- a parameter-smuggling asymmetry versus downstream/naive Content-Disposition parsers.

Reproduced: `Content-Disposition: form-data; name=ab"cd; filename="ev.txt"` yields a form *field* literally named `ab"cd; filename="ev.txt` containing the file bytes (no File entry); `name="a; filename="x.txt"` likewise demotes the upload to a field where the old code returned file `x.txt`.

File: src/cpp/core/http/Util.cpp [hunk 7]
```diff
@@ -121,117 +125,335 @@ void buildQueryString(const Fields& fields, std::string* pQueryString)
...
-   
-   // extract the fields
-   size_t beginBoundaryLoc = body.find(boundary);
-   size_t endBoundaryLoc = body.find("\r\n" + boundary,
-                                     beginBoundaryLoc+boundary.size());
-   while (endBoundaryLoc != std::string::npos)
+
+   // Per RFC 1341, multipart-body ends immediately after the `--` with no CRLF necessary.
+   size_t terminatorPos = body.find("\r\n--" + boundary + "--");
+   if (!terminatorPos || !body.size())
```
Comment (kevinushey) [inline src/cpp/core/http/Util.cpp]: **`if (!terminatorPos || !body.size())` is booby-trapped, and the shared warning mislabels valid input.**

- `!terminatorPos` reads as "not found" but actually means "found at index 0"; the intended "not found" sentinel is `std::string::npos` (handled below at 370).
- The `!body.size()` arm looks redundant but is the only guard for the `&*body.begin()` deref at line 373 -- easy for a later "simplification" to drop and reintroduce UB on empty bodies.
- The shared `Invalid multipart/form-data: no sections` warning mislabels a well-formed zero-part multipart as invalid, and it now fires per-request on the unauthenticated sign-in path (`ServerPAMAuth::doSignIn` calls `formFieldValue` before validation), where the old parser was silent.

Clearer as `if (body.empty() || terminatorPos == 0)` with distinct messages (and consider debug-level on the pre-auth path).

File: src/cpp/core/http/Util.cpp [hunk 8]
```diff
@@ -121,117 +125,335 @@ void buildQueryString(const Fields& fields, std::string* pQueryString)
...
+      auto filenameIter = cDisp.params.find("filename");
+      if (filenameIter != cDisp.params.end() && pFiles->find(formName) != pFiles->end())
+      {
+         // If we've already processed a file upload for a given form field,
+         // keep the first file and ignore the rest without allocating any
+         // additional memory.
+         continue;
+      }
+
+      std::string formValue(partRange.size(), '\0');
```
Comment (kevinushey) [inline src/cpp/core/http/Util.cpp:452]: **Part-content read does ~3 memory passes where 1 suffices.**

`std::string(partRange.size(), '\0')` zero-fills the entire part (headers included), then `filtering_istream::read` copies through a 4 KB internal buffer (`indirect_streambuf` only overrides `underflow`, so bytes go body -> 4 KB buffer -> `formValue` in ~4 KB chunks). The line-390 comment "wrap in non-copying stream" isn't true for the read path.

A 100 MB upload (`RequestParser::maxContentLength`) pays a ~100 MB memset plus ~200 MB of buffered copy traffic (~25k virtual `underflow` calls) on the request thread versus a single exact-size copy. Cheaper: keep the stream only for `parseHeaders` (~100 bytes), then construct `formValue` directly from the contiguous `partRange` tail after the `\r\n\r\n` separator -- which also fixes the misleading comment.

File: src/cpp/core/http/Util.cpp [hunk 9]
```diff
@@ -17,17 +17,21 @@
...
+#include <cctype>
 #include <ios>
 #include <iostream>
 #include <sstream>
+#include <string_view>
 #include <algorithm>
 
 #include <boost/asio.hpp>
 #include <boost/tokenizer.hpp>
 #include <boost/algorithm/string/classification.hpp>
```
Comment (kevinushey) [inline src/cpp/core/http/Util.cpp:30]: **Two includes now rely on fragile transitive chains.**

This backport drops `#include <boost/algorithm/string/case_conv.hpp>` (kept by both the pro original and the OSS base) while `boost::algorithm::to_lower`/`to_lower_copy` are still used at lines 308/405/873; and it adds new `std::uint8_t` uses (293/295) without `<cstdint>`. Both currently compile only via transitive includes (gregorian.hpp, asio.hpp). Any boost upgrade or include cleanup that trims those chains breaks this TU with confusing errors. Restoring `case_conv.hpp` (matching pro) and adding `<cstdint>` also reduces future merge divergence from pro.

File: src/cpp/core/http/Util.cpp [hunk 10]
```diff
@@ -121,117 +125,335 @@ void buildQueryString(const Fields& fields, std::string* pQueryString)
...
+            inQuotes = !inQuotes;
+         }
+         else if (!inQuotes && *iter == ';')
+         {
+            return boost::iterator_range<ITER>(iter, iter + 1);
+         }
+      }
+      return boost::iterator_range<ITER>(end, end);
+   }
+} semiFinder;
```
Comment (kevinushey) [inline src/cpp/core/http/Util.cpp:187]: **Hygiene roll-up on the new parser code.**

- `} semiFinder;` defines a mutable global with external linkage in a named namespace (ODR/link hazard). Prefer `const` in an anonymous namespace, or just pass `SemiFinder{}` at the single call site.
- Comment typo "permissinve", and two citations of RFC 1341 -- the multipart grammar you're implementing lives in RFC 2046 (1341 was obsoleted in 1992), so a maintainer verifying the terminator behavior is pointed at the wrong spec.
- `HeaderParams::parse`'s `&*headerRange.begin()` is UB for a future empty-range caller; a `string_view` parameter would remove both the template and the trap.
- Style: the new structs' same-line opening braces (`operator()(...) const {`, `enum State {`) don't match the Allman style the repo's clang-format uses for `src/cpp`.

File: src/cpp/core/http/RequestParserTests.cpp [hunk 11]
```diff
@@ -209,75 +244,177 @@ TEST(HttpTest, ComplicatedFormParsingWorksOneByteAtATime)
...
 
-   std::string bodyStr;
-   std::string requestStr = complexRequest(fileBytes, &bodyStr);
-   Request request;
+   RequestParser::status status;
+   size_t blockSize = 0;
+   do
+   {
+      blockSize = rand() % 8192 + 1;
+      status = form.parseBytes(blockSize);
```
Comment (kevinushey) [inline src/cpp/core/http/RequestParserTests.cpp:288]: **Random-chunk test violates the parser's resume contract (latent, confirmed empirically).**

After `headers_parsed`, `parseBytes` correctly re-passes the same `parseIter`, but this loop draws a *fresh* `rand() % 8192 + 1` count. The parser then does `begin += bufferPos_` (121 for this request); if the new count < 121, `begin` lands past `end`, `copyRangeToBuffer`'s `std::distance` goes negative, `neededCapacity` wraps to ~2^64, and `formBuffer_.reserve()` throws `std::length_error`.

Confirmed: `parse(begin, begin+424)` -> `headers_parsed`, then `parse(begin, begin+50)` throws `std::length_error` (`begin+121` is the exact safe threshold). Unseeded `rand()` currently draws 424 then 6898 on macOS -- safe by luck; ~1.5% of the draw space is dangerous, and any libc change, `srand()`, or test-order shift turns this into a baffling per-platform failure. Fix by re-passing the previous block on `headers_parsed` (as the old test did), or clamping `bufferPos_` in `parse()`.

File: src/cpp/core/http/RequestParserTests.cpp [hunk 12]
```diff
@@ -76,107 +52,166 @@ core::Result<std::string> generateRandomBytes()
...
+   }
+
+   RequestParser::status parse()
+   {
+      const char* begin = requestStr.c_str();
+      const char* end = begin + requestStr.size();
+      return parser.parse(request, begin, end);
+   }
+
+   RequestParser::status parseBytes(int count)
```
Comment (kevinushey) [inline src/cpp/core/http/RequestParserTests.cpp:138]: **Parse loops lack a progress guarantee and an exact-completion check.**

`headers_parsed` is accepted forever with `parseIter` frozen, so a resume regression that keeps returning `headers_parsed` spins this `do`/`while` until CI timeout (a hang, not a red assertion -- `eof()` never becomes true and the ASSERT permits the status). And because the loops break on `form_complete` wherever it occurs while `parseIter` is also frozen on completion, a parser that returns `form_complete` early with unconsumed bytes is undetectable; `validationError` is never checked in the three `Complicated*` tests. The old tests asserted completion happened exactly at the final byte -- worth preserving that invariant.

File: src/cpp/core/http/RequestParserTests.cpp [hunk 13]
```diff
@@ -76,107 +52,166 @@ core::Result<std::string> generateRandomBytes()
...
+      return !parseIter || parseIter >= endIter;
+   }
+
+   std::string boundary;
+   std::string expectedData;
+   std::string requestStr;
+   std::string buffer;
+   Error validationError;
+   Request request;
+   FormHandler handler;
```
Comment (kevinushey) [inline src/cpp/core/http/RequestParserTests.cpp]: **Dead member, and a copyable tester holding a `this`-capturing lambda.**

`FormHandler handler;` here is unused. Separately, `FormTester` is copyable while its parser holds a lambda capturing `this`, so a copy (e.g. stored by value in a container) mutates/dangles the original -- worth deleting the copy ops to turn a silent runtime bug into a compile error.

File: src/cpp/core/http/RequestParserTests.cpp [hunk 14]
```diff
@@ -209,75 +244,177 @@ TEST(HttpTest, ComplicatedFormParsingWorksOneByteAtATime)
...
+   EXPECT_EQ(form.request.formFieldValue("field1"), "value1");
+
+   File file = form.request.uploadedFile("field2");
+   EXPECT_FALSE(file.empty());
+   EXPECT_EQ(file.name, "example.txt");
+   EXPECT_EQ(file.contentType, "text/plain");
+   EXPECT_EQ(file.contents, "This is a simple text file");
+}
+
+TEST(HttpTest, FormParsingIgnoresEpilogue)
```
Comment (kevinushey) [inline src/cpp/core/http/RequestParserTests.cpp:350]: **Coverage gaps on exactly the new code paths.**

- `FormParsingIgnoresEpilogue` strips the delimiter from its epilogue, so no implementation could ever parse `field3` -- the assertion is vacuous, and `field2.contents` is never checked.
- The edge-case/padding tests assert only `file.name`, never `contents` or field counts.
- Untested new branches: missing-terminator best-effort, the `!terminatorPos` guard, per-part skips (no Content-Disposition, non-form-data, missing name, `values.empty()` guard), duplicate-file keep-first, and boundary-prefix rejection (`--BB` vs `--BBB`). A regression that stops ignoring the epilogue, shifts content by the padding width, inverts a skip, or flips keep-first to keep-last would pass the suite unchanged.

**Comment by kevinushey** [comment]
Automated review pass on the multipart rewrite (max effort).

I compiled the new `parseMultipartForm` into a standalone harness and compared old-vs-new behavior on ~60 hostile inputs (ASan-clean); the inline comments below marked "confirmed by run" are reproductions against that harness, not static guesses.

Themes: a few of the new strictness/leniency choices silently drop or corrupt parts on adversarial-but-not-impossible input (the `find()` pre-scan, padded close-delimiter, quote toggling, preamble handling), and the quoted-param backslash-unescaping is a real regression for browser-sent filenames. None of these block the DoS fix; they're worth a look before merge.

One finding doesn't map to a changed line, so noting it here: the Files-pane `/upload` streaming parser in `src/cpp/session/modules/SessionFiles.cpp` (`getBoundary`/`parseContentBuffer`, around line 1085) is a verbatim copy of the *old* grammar this PR deletes -- case-sensitive unquoted `boundary=`, quote-blind filename parsing, and a third terminator grammar requiring a trailing CRLF. So the inputs this PR's new tests legitimize (e.g. `Content-Type: multipart/form-data; Boundary="boundary"`) parse via `Request::formFields()`/`uploadedFile()` but fail with "Invalid form data received" when POSTed to `/upload`. Worth a follow-up to route both paths through one boundary/disposition parser.

Details inline.

File: src/cpp/core/http/Util.cpp (hunk 9 above)
Comment (adamhigerd) [reply src/cpp/core/http/Util.cpp:30]: Ack! Bad merge resolution! I thought I was in the clear when the build succeeded and the tests passed but I overlooked that particular lost `#include`. Thanks.

File: src/cpp/core/http/Util.cpp (hunk 2 above)
Comment (adamhigerd) [reply src/cpp/core/http/Util.cpp:349]: **Rejected.**

RFC 7578 says no such thing; that admonition is against assuming that the filename is a legal filename. It also says that agents MAY use percent-encoding for non-US-ASCII characters.

The normative specification is RFC 6266, which states that the `filename` parameter follows the grammar `token | quoted-string`. Since this code path is only used when the parameter begins and ends with `"`, it is a `quoted-string`, which treats `\` as an unconditional escape sequence.

File: src/cpp/core/http/Util.cpp (hunk 7 above)
Comment (adamhigerd) [reply src/cpp/core/http/Util.cpp]: Ehhh... that's quite nitpicky, but... I can't disagree with using `== 0` here since I'm making changes anyway.

I reject the remaining points:
* `!body.size()` doesn't look redundant at all.
* RFC 2046 defines multipart entities as "one or more different sets of data", meaning a zero-part multipart violates the specification.
* Even if it was valid, I frankly don't care, as it still returns an empty form and it produces a harmless warning in a code path that's just going to throw another error anyway.

File: src/cpp/core/http/Util.cpp (hunk 1 above)
Comment (adamhigerd) [reply src/cpp/core/http/Util.cpp:387]: Ugh. While 2 and 3 are invalid (a delimiter by definition cannot be contained within the preamble, and is _supposed_ to truncate the view and drop every later part), 1 is... annoyingly meaningful.

I might punt on fixing it for now, though, because it's release day and this is supposed to be a backport.

File: src/cpp/core/http/Util.cpp (hunk 4 above)
Comment (adamhigerd) [reply src/cpp/core/http/Util.cpp:396]: **Rejected.** The choice to be lenient doesn't excuse the sender's obligation to send correct data. Garbage in, garbage out.

File: src/cpp/core/http/Util.cpp (hunk 5 above)
Comment (adamhigerd) [reply src/cpp/core/http/Util.cpp:409]: **Rejected.** RFC 2046 defines the `\r\n` as normatively required.

File: src/cpp/core/http/Util.cpp (hunk 2 above)
Comment (kevinushey) [reply src/cpp/core/http/Util.cpp:349]: Fair point on the RFC 7578 citation -- s4.2's admonition (via RFC 2183 s2.3) is about filename legality and paths, not quoted-string escaping. I'll take that back.

RFC 6266 can't be the normative spec here, though; it disclaims exactly this case in Section 1:

> Note: This document does not apply to Content-Disposition header fields appearing in payload bodies transmitted over HTTP, such as when using the media type "multipart/form-data" ([RFC2388]).

The senders we're parsing are governed by the WHATWG HTML multipart/form-data encoding algorithm, which requires that field names and filenames be escaped by replacing LF with `%0A`, CR with `%0D`, and `"` with `%22`, and then says "The user agent must not perform any other escapes":

https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#multipart-form-data

So every conformant browser sends literal backslashes raw, and unescaping them corrupts real uploads -- that's the behavior reproduced above, and RStudio's own frontend is such a browser. The WHATWG parsing counterpart (now in the Fetch spec for Response.formData()) likewise gives backslash no special treatment inside quoted values.

Concretely: keep quoted-pair unescaping where it's legitimate -- the HTTP-level Content-Type/boundary parse (RFC 9110 quoted-string) -- but for the part-level Content-Disposition name/filename params, strip the outer quotes and take the bytes as-is. The FormParsing test expectation of `semi\colon;and"quote.txt` from the `\"`-escaped wire form would change accordingly; that wire form is one no browser can produce, since quotes always arrive as `%22`. Same change would be needed in the pro source PR.

Comment (kevinushey) [reply src/cpp/core/http/Util.cpp:349]: A small example to demonstrate:

```
cd "$(mktemp -d)" && printf hi > 'a\b.txt' && printf hi > 'a"b.txt' && { python3 - <<'EOF' &
import socket
s = socket.socket()
s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
s.bind(("127.0.0.1", 8998))
s.listen(1)
c, _ = s.accept()
c.settimeout(1)
data = b""
try:
    while True:
        chunk = c.recv(65536)
        if not chunk:
            break
        data += chunk
except socket.timeout:
    pass
c.sendall(b"HTTP/1.1 200 OK\r\nContent-Length: 0\r\n\r\n")
for line in data.split(b"\r\n"):
    if line.lower().startswith(b"content-disposition"):
        print(line.decode("latin1"))
EOF
sleep 1; curl -s -F 'upload=@a\b.txt' -F 'upload2=@a"b.txt' -o /dev/null http://127.0.0.1:8998/; wait; }
```

which gives me, when run

```
$ bash ./script.sh
Content-Disposition: form-data; name="upload"; filename="a\b.txt"
Content-Disposition: form-data; name="upload2"; filename="a%22b.txt"
```

File: src/cpp/core/http/Util.cpp (hunk 8 above)
Comment (adamhigerd) [reply src/cpp/core/http/Util.cpp:452]: **Rejected.**

The entire upload being stored in RAM before parsing is an existing known issue. This implementation improves substantially on the status quo, and doing better would be too invasive for this specific task -- better to handle it separately.

"Non-copying" here means that it doesn't create a duplicated copy of the entire payload, so the total RAM usage is lower. Yes, obviously the stream is going to be copying intermediate portions, but this isn't what we're worried about.

File: src/cpp/core/http/RequestParserTests.cpp (hunk 12 above)
Comment (adamhigerd) [reply src/cpp/core/http/RequestParserTests.cpp:138]: **Rejected.**

No need to assert that completion is at the final byte; indeed, we EXPECT to stop parsing before then.

It would be valid that this loop could get stuck, but a timeout is still a test failure. Less than ideal, but not worth complicating the test harness to avoid it when it shouldn't ever happen in practice anyway.

File: src/cpp/core/http/RequestParserTests.cpp (hunk 13 above)
Comment (adamhigerd) [reply src/cpp/core/http/RequestParserTests.cpp]: Eh... almost decided to ignore this one but it's trivial to fix.

File: src/cpp/core/http/RequestParserTests.cpp (hunk 11 above)
Comment (adamhigerd) [reply src/cpp/core/http/RequestParserTests.cpp:288]: **Known issue.** Deferring.

File: src/cpp/core/http/RequestParserTests.cpp (hunk 14 above)
Comment (adamhigerd) [reply src/cpp/core/http/RequestParserTests.cpp:350]: **Rejected.**

* "no implementation could ever parse `field3`" is precisely what that assertion is testing, which invalidates the comment that it would "pass the suite unchanged".
* No need to re-assert what other tests have already covered. The additional tests are only covering changed behavior.
* Intentionally didn't put in tests for the lenient behavior; it's not normative and can change without impacting overall correctness.
* Boundary-prefix rejection and duplicate-file keep-first are vaguely relevant but not something I'm going to do right now. Consider it **deferred**.

File: src/cpp/core/http/Util.cpp (hunk 2 above)
Comment (adamhigerd) [reply src/cpp/core/http/Util.cpp:349]: Oh, bother.

In that case, **deferred.**

File: src/cpp/core/http/Util.cpp (hunk 1 above)
Comment (adamhigerd) [reply src/cpp/core/http/Util.cpp:387]: **Deferred.**

File: src/cpp/core/http/Util.cpp (hunk 2 above)
Comment (adamhigerd) [reply src/cpp/core/http/Util.cpp:349]: ... Actually? No, I changed my mind.

**Rejected.**

Consumers of this functionality are either going to ignore the literal filename, or will need to handle invalid filenames somehow anyway. The complexity necessary to do this "right" isn't worth the effort.

**Comment by adamhigerd** [comment]
`SessionFiles.cpp` fix is also a **known issue**, likewise deferred.

File: src/cpp/core/http/Util.cpp (hunk 1 above)
Comment (adamhigerd) [reply src/cpp/core/http/Util.cpp:387]: Discovered a much simpler fix than I feared -- turns out the only reason it's a problem at all is because I mistakenly cut off the payload at the _start_ of the terminator instead of the _end_ of it.

Still deferred, though. Not going to fix that in this PR, because no real-world client will ever generate an empty multipart section.

Comment (adamhigerd) [reply src/cpp/core/http/Util.cpp:387]: Actually... while I was writing up the follow-up ticket, I discovered that this is wrong.

> The boundary may be followed by zero or more characters of
> linear whitespace. It is then terminated by either another CRLF and
> the header fields for the next part, or by two CRLFs, in which case
> there are no header fields for the next part. 

It's not _possible_ to trigger this failure case with a compliant client implementation, so I hereby **reject** this case as GIGO.

(That said, I'm still going to make the change I described, because it was supposed to be written that way the whole time.)

**Comment by adamhigerd** [comment]
Follow-up: https://github.com/rstudio/rstudio/issues/18727
